### PR TITLE
B 17615 weight ticket designator for payment request uploads

### PR DESCRIPTION
**Pull request by danieljordan-caci**
## Summary

Added a requirement of the Prime to add a parameter of `isWeightTicket` to a POST request to the `/createUpload` endpoint. 

This required adding a migration file to add a column, `isWeightTicket` to the `proof_of_service_docs` table. Made the default value of `FALSE` to avoid any `NULL` panics.

Updated the Go Struct for `ProofOfServiceDocs` model to include the `isWeightTicket` boolean value.

Updated the payload that is returned for an upload if the upload is:
1. A proof of service doc
2. A proof of service doc that is a weight ticket
This adds an `isWeightTicket` key to the upload array that is used by the UI when reviewing payment requests and their respective documents.

*NOTE* This story did not have criteria to update the Prime simulator, so when using the Prime Sim you will be unable to test this or add a `isWeightTicket` value to the upload.

In order to see this change in the frontend, I also had to update the `model_to_payload.go` file for the GHC API and allow including the parameter in a `GET` request for payment requests.

*DON'T FREAK OUT* by the number of "Files changed" in this request. When changing any property in `swagger-def`, it changes several auto-generated files. You good. We good. It's all good.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Log into MilMove as the Prime
2. Using Postman, send a POST request to `/createUpload` on an existing payment request. Ensure your body is in `formData` as the `Content-Type` and send it.
```
{
   "file": "filePath",
   "isWeightTicket": true
}
```
3. Sign in as an office user and perform a `GET` request to `/payment-requests/{paymentRequestID}`
4. Confirm you can see the `isWeightTicket` parameter for the `proofOfServiceDoc` and `upload` you added to the payment request

### CODE REVIEW

**Comment by cameroncaci** [comment]
Since this was drafted for a while it won't appear at the top. Pinging everyone for awareness: @deandreJones @TevinAdams @traskowskycaci

File: migrations/app/schema/20231109142950_add-weight-ticket-boolean-to-uploads-table.up.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+-- adding column to uploads table so we can determine if an upload is a weight ticket or not
+-- this will be a boolean data type
+ALTER TABLE proof_of_service_docs
+ADD COLUMN is_weight_ticket boolean DEFAULT false;
```
Comment (cameroncaci) [inline migrations/app/schema/20231109142950_add-weight-ticket-boolean-to-uploads-table.up.sql:4]: Almost forgot, this needs a not null attribute

**Review by cameroncaci — CHANGES_REQUESTED** [review]

File: migrations/app/schema/20231109142950_add-weight-ticket-boolean-to-uploads-table.up.sql (hunk 1 above)
Comment (danieljordan-caci) [reply migrations/app/schema/20231109142950_add-weight-ticket-boolean-to-uploads-table.up.sql]: Corrected

**Review by traskowskycaci — APPROVED** [review]
Tested with swagger and postman, looks good!

File: pkg/handlers/ghcapi/internal/payloads/model_to_payload.go [hunk 2]
```diff
@@ -1423,6 +1423,28 @@ func Upload(storer storage.FileStorer, upload models.Upload, url string) *ghcmes
...
+		IsWeightTicket: isWeightTicket,
+	}
+	tags, err := storer.Tags(upload.StorageKey)
+	if err != nil || len(tags) == 0 {
+		uploadPayload.Status = "PROCESSING"
+	} else {
+		uploadPayload.Status = tags["av-status"]
+	}
+	return uploadPayload
+}
```
Comment (cameroncaci) [inline pkg/handlers/ghcapi/internal/payloads/model_to_payload.go:1446]: Question,
Why is the error handlign here only set to `PROCESSING` and not returning an actual error in the func?

Comment (danieljordan-caci) [reply pkg/handlers/ghcapi/internal/payloads/model_to_payload.go:1446]: No idea, this was an implementation in the other function I copied this from. Code from Truss days.

File: pkg/handlers/ghcapi/internal/payloads/model_to_payload.go [hunk 3]
```diff
@@ -1433,12 +1455,29 @@ func ProofOfServiceDoc(proofOfService models.ProofOfServiceDoc, storer storage.F
...
+			weightTicket := proofOfService.IsWeightTicket
+			if *weightTicket {
+				uploads[i] = WeightTicketUpload(storer, primeUpload.Upload, url, *proofOfService.IsWeightTicket)
+			} else {
+				uploads[i] = Upload(storer, primeUpload.Upload, url)
+			}
 		}
 	}
 
+	// checking if isWeightTicket is NULL to avoid panic
```
Comment (cameroncaci) [inline pkg/handlers/ghcapi/internal/payloads/model_to_payload.go:1468]: Where did this panic by the way? With our column value set to NOT NULL now I'm not sure if we need this anymore

File: pkg/factory/proof_of_service_doc_factory.go [hunk 4]
```diff
@@ -28,10 +28,12 @@ func BuildProofOfServiceDoc(db *pop.Connection, customs []Customization, traits
 	}
 
 	paymentRequest := BuildPaymentRequest(db, customs, traits)
+	isWeightTicket := false
 
 	ProofOfServiceDoc := models.ProofOfServiceDoc{
 		PaymentRequest:   paymentRequest,
 		PaymentRequestID: paymentRequest.ID,
+		IsWeightTicket:   &isWeightTicket,
```
Comment (cameroncaci) [inline pkg/factory/proof_of_service_doc_factory.go:36]: Is this still needed to be a pointer if `NOT NULL`?

File: pkg/models/proof_of_service_doc.go [hunk 5]
```diff
@@ -15,6 +15,7 @@ type ProofOfServiceDoc struct {
 	PaymentRequestID uuid.UUID `json:"payment_request_id" db:"payment_request_id"`
 	CreatedAt        time.Time `db:"created_at"`
 	UpdatedAt        time.Time `db:"updated_at"`
+	IsWeightTicket   *bool     `db:"is_weight_ticket"`
```
Comment (cameroncaci) [inline pkg/models/proof_of_service_doc.go:18]: Just another quick pointer check if the DB column has `NOT NULL`

Comment (danieljordan-caci) [reply pkg/models/proof_of_service_doc.go:18]: FIXED YOU CRAZY KID

File: pkg/factory/proof_of_service_doc_factory.go (hunk 4 above)
Comment (danieljordan-caci) [reply pkg/factory/proof_of_service_doc_factory.go:36]: This has been signed, sealed, and delivered

File: pkg/handlers/ghcapi/internal/payloads/model_to_payload.go (hunk 3 above)
Comment (danieljordan-caci) [reply pkg/handlers/ghcapi/internal/payloads/model_to_payload.go:1468]: I panicked and fixed this for you

**Review by cameroncaci — APPROVED** [review]

**Comment by cameroncaci** [comment]
Tested successfully locally!

**Comment by deandreJones** [comment]
conflicts

**Comment by danieljordan-caci** [comment]
> conflicts

Resolved

**Pull request merged**